### PR TITLE
predict function

### PR DESCRIPTION
**Pull request by cchallu**

### CODE REVIEW

File: neuralforecast/models/nbeats/nbeats.py [hunk 1]
```diff
@@ -797,3 +797,121 @@ def forecast(self: NBEATS, Y_df: pd.DataFrame, X_df: pd.DataFrame = None, S_df:
     forecast_df['y'] = forecast.flatten()
 
     return forecast_df
+
+
+# Cell
+@patch
+def predict(self: NBEATS, Y_df: pd.DataFrame, X_df: pd.DataFrame = None, S_df: pd.DataFrame = None,
```
Comment (FedericoGarza) [inline neuralforecast/models/nbeats/nbeats.py:804]: Probably `predict_in_sample` would be a better name? I think the `predict` method should be the same as the `forecast` method if we want to include it.

File: neuralforecast/models/nhits/nhits.py [hunk 2]
```diff
@@ -787,3 +906,125 @@ def forecast(self: NHITS, Y_df: pd.DataFrame, X_df: pd.DataFrame = None, S_df: p
...
+    fcds_column = fcds.repeat(model.n_time_out)
+
+    # Output DataFrame
+    Y_hat_df = pd.DataFrame({'unique_id':unique_ids_column,
+                             'forecast_creation_ds': fcds_column,
+                             'ds':ds_rolled.flatten(),
+                             'y': y_true.flatten(),
+                             'y_hat':y_hat.flatten()})
+
+    return Y_hat_df
```
Comment (FedericoGarza) [inline neuralforecast/models/nhits/nhits.py:1030]: The code is the same as the `predict` method of `NBEATS`. Maybe we should abstract it.

File: neuralforecast/data/tsdataset.py [hunk 3]
```diff
@@ -642,8 +648,14 @@ def __init__(self,
         # WindowsDataset parameters
         self.windows_size = self.input_size + self.output_size
         self.padding = (self.input_size, self.output_size)
+
+        if len(self.ds) == (self.max_len+self.input_size+self.output_size):
+            print('WARNING: THIS DATA LOADER ASSUMES ALL TIME SERIES END IN THE SAME DATESTAMP, CAN CAUSE MISMATCH IN DATES WHILE FORECASTING!!!')
+            time.sleep(5)
```
Comment (FedericoGarza) [inline neuralforecast/data/tsdataset.py:654]: Why a `time.sleep` here?

**Review by FedericoGarza — CHANGES_REQUESTED** [review]

File: neuralforecast/models/nbeats/nbeats.py (hunk 1 above)
Comment (cchallu) [reply neuralforecast/models/nbeats/nbeats.py:804]: lets use fit predict, its more common

**Review by FedericoGarza — APPROVED** [review]

**Review by FedericoGarza — CHANGES_REQUESTED** [review]
.

File: neuralforecast/models/nhits/nhits.py [hunk 4]
```diff
@@ -719,6 +719,125 @@ def configure_optimizers(self):
...
+
+    # Extend ds vector to match padded ts_tensor
+    timedelta = dataset.ds[1] -dataset.ds[0]
+    start_date = dataset.ds[0] - dataset.input_size*timedelta
+    end_date = dataset.ds[-1] + dataset.output_size*timedelta
+    ds = pd.date_range(start_date, end_date, freq=dataset.frequency)
+
+    # Tensor with sample_mask
+    tensor = dataset.ts_tensor[:,dataset.t_cols.index('sample_mask'),:]
+    padding = (model.n_time_in, model.n_time_out)
```
Comment (FedericoGarza) [inline neuralforecast/models/nhits/nhits.py:796]: `model` here is not defined, maybe you meant `self`?

File: neuralforecast/models/nbeats/nbeats.py [hunk 5]
```diff
@@ -797,3 +797,121 @@ def forecast(self: NBEATS, Y_df: pd.DataFrame, X_df: pd.DataFrame = None, S_df:
...
+
+    # Extend ds vector to match padded ts_tensor
+    timedelta = dataset.ds[1] -dataset.ds[0]
+    start_date = dataset.ds[0] - dataset.input_size*timedelta
+    end_date = dataset.ds[-1] + dataset.output_size*timedelta
+    ds = pd.date_range(start_date, end_date, freq=dataset.frequency)
+
+    # Tensor with sample_mask
+    tensor = dataset.ts_tensor[:,dataset.t_cols.index('sample_mask'),:]
+    padding = (model.n_time_in, model.n_time_out)
```
Comment (FedericoGarza) [inline neuralforecast/models/nbeats/nbeats.py:874]: `model` is not defined, maybe `self`?

File: neuralforecast/models/nbeats/nbeats.py [hunk 6]
```diff
@@ -797,3 +797,121 @@ def forecast(self: NBEATS, Y_df: pd.DataFrame, X_df: pd.DataFrame = None, S_df:
...
+    padder = t.nn.ConstantPad1d(padding=padding, value=0)
+    tensor = padder(tensor)
+
+    # Add dates indexes to tensor
+    ds_idxs = t.tile(t.arange(start=0, end=len(ds)),dims=[len(tensor), 1])
+    tensor = t.cat((tensor[:,None,:],ds_idxs[:,None,:]),dim=1)
+
+    # Rolling window of sampling mask and dates indexes
+    tensor = tensor.unfold(dimension=-1,
+                            size=model.n_time_in+model.n_time_out,
```
Comment (FedericoGarza) [inline neuralforecast/models/nbeats/nbeats.py:884]: same problem here

File: neuralforecast/models/nbeats/nbeats.py [hunk 7]
```diff
@@ -797,3 +797,121 @@ def forecast(self: NBEATS, Y_df: pd.DataFrame, X_df: pd.DataFrame = None, S_df:
...
+
+    # Split rolled tensor in sample_mask and ds
+    ds_rolled = ds[tensor[:,1,:,-model.n_time_out-1:].numpy().astype(int)] # -model.n_time_out-1 to add forecast creation date
+    sample_mask_rolled = tensor[:,0,:,-model.n_time_out:]
+
+    # Sampling condition, ALL ds in the forecast window should be sampleaple
+    sample_condition = (sample_mask_rolled.sum(dim=2)==model.n_time_out)
+
+    # Filter rolled ds with sample_condition
+    ds_rolled = ds_rolled.reshape(-1, model.n_time_out+1)
```
Comment (FedericoGarza) [inline neuralforecast/models/nbeats/nbeats.py:895]: references to `model` but it is not defined

**Review by FedericoGarza — CHANGES_REQUESTED** [review]
Check `model` calls in the `predict` method, it is not defined. Maybe `self`?

**Pull request closed without merging**